pkg/reconciler/brokercell/brokercell.go

```diff
@@ -95,13 +94,22 @@ var _ bcreconciler.Interface = (*Reconciler)(nil)
 func (r *Reconciler) ReconcileKind(ctx context.Context, bc *intv1alpha1.BrokerCell) pkgreconciler.Event {
 	bc.Status.InitializeConditions()
 
-	// Reconcile ingress deployment and service
+	// Reconcile ingress deployment, HPA and service.
 	ingressArgs := r.makeIngressArgs(bc)
-	if _, err := r.deploymentRec.ReconcileDeployment(bc, resources.MakeIngressDeployment(ingressArgs)); err != nil {
+	ind, err := r.deploymentRec.ReconcileDeployment(bc, resources.MakeIngressDeployment(ingressArgs))
+	if err != nil {
 		logging.FromContext(ctx).Error("Failed to reconcile ingress deployment for \"%s/%s\": %v", zap.Any("namespace", bc.Namespace), zap.Any("name", bc.Name), zap.Error(err))
 		bc.Status.MarkIngressFailed("IngressDeploymentFailed", "Failed to reconcile ingress deployment: %v", err)
 		return err
 	}
+
+	ingressHPA := resources.MakeHorizontalPodAutoscaler(ind, r.makeIngressHPA(bc))
+	if err := r.reconcileAutoscaling(ctx, bc, ingressHPA); err != nil {
+		logging.FromContext(ctx).Error("Failed to reconcile ingress HPA for \"%s/%s\": %v", zap.Any("namespace", bc.Namespace), zap.Any("name", bc.Name), zap.Error(err))
+		bc.Status.MarkIngressFailed("HorizontalPodAutoscalerFailed", "Failed to reconcile ingress HorizontalPodAutoscaler: %v", err)
```

Comment:
I don't see a very compelling reason to not treat HPA as part of the component unit (e.g. ingress unit). If HPA fails, it could have a real impact on the ingress availability under some circumstances. That's why I prefer my current approach.

If we really want to treat HPA as a 2nd party citizen, we can probably add a *single* condition type for HPA and indicate which HPA has failed in the message.

Reply:
I am fine with the current approach. I was debating whether we should treat HPA as 2nd party citizen or even whether we should return an error if HPA fails since very likely the data plane still works without the scaling. But I guess it's too tricky and complicated if we that.